src/render.c

```diff
@@ -1,22 +1,296 @@
 #include <config.h>
 
+#include <math.h>
 #include <time.h>
 #include <signal.h>
 #include <curses.h>
 #include <stdlib.h>
+#include <string.h>
+#include <term.h>
 #include "render.h"
 #include "pipe.h"
+#include "util.h"
 
 #define NS 1000000000L //1ns
 
-void init_colours(void){
-    //Initialise colour pairs if we can.
-    if(has_colors()){
-        start_color();
-        for(short i=1; i < COLORS; i++){
-            init_pair(i, i, COLOR_BLACK);
+#define ESCAPE_CODE_SIZE 256
+
+// Number of reserved pairs in indirect mode (just colour pair 0)
+#define RESERVED_INDIRECT_PAIRS 1
+
+static int hsl2rgb(float hue, float sat, float light);
+static bool have_direct_colours(void);
+static int set_color_pair_direct(int color_index, int color);
+static int set_color_pair_indirect(int color_index, int color);
+static int set_pair(int pair_index, int fg, int bg);
+static int palette_size(void);
+
+static int query_terminal(const char *escape, char *buffer, int bufsz);
+
+// Cribbed straight from Wikipedia :)
+int hsl2rgb(float hue, float sat, float light) {
+    float chroma = (1.0f - fabs(2*light - 1)) * sat;
+    float hue_segment = hue / 60;
+    float cpt2 = chroma * (1 - fabs(fmod(hue_segment, 2) - 1));
+    float r, g, b;
+    switch((int)hue_segment){
+        case 0: r = chroma; g = cpt2  ; b = 0     ; break;
+        case 1: r = cpt2  ; g = chroma; b = 0     ; break;
+        case 2: r = 0     ; g = chroma; b = cpt2  ; break;
+        case 3: r = 0     ; g = cpt2  ; b = chroma; break;
+        case 4: r = cpt2  ; g = 0     ; b = chroma; break;
+        case 5: r = chroma; g = 0     ; b = cpt2  ; break;
+        default: r = 0; g = 0; b = 0;
+    }
+    float offset = light - chroma/2;
+    r += offset;
+    g += offset;
+    b += offset;
+
+    int ir = (int)(r * 255);
+    int ig = (int)(g * 255);
+    int ib = (int)(b * 255);
+    return (ir << 16) | (ig << 8) | ib;
+}
+
+// Check whether this terminal has the RGB capability (direct colors).
+// This is adapted from `parse_rgb.h` header with the ncurses test programs.
+bool have_direct_colours(void) {
+    if (tigetflag("RGB") > 0) {
+        return true;
+    } else if (tigetnum("RGB") > 0) {
+        return true;
+    } else {
+        char *strval = tigetstr("RGB");
+        if(strval != (char *) -1 && strval)
+            return true;
+    }
+    return false;
+}
+
+/// Returns the effective size of the palette that we can use.
+int palette_size(void) {
+    bool direct = have_direct_colours();
+
+    // On terminals that support direct colour, COLORS can be greater than
+    // COLOR_PAIRS. We can only display `COLOR_PAIRS` different colours,
+    // though, so that is the size of our palette.
+    //
+    // Note that for terminals that do not support direct colour, we reduce the
+    // possible number of colours by 8, because we should not alter the default
+    // 8 colours.
+    int max_pipe_colors = min(COLORS, COLOR_PAIRS);
+
+#if !HAVE_ALLOC_PAIR
```

Comment:
Here is the output I got as of 4b143e5:

```c
src/render.c: In function 'palette_size':
src/render.c:84:6: warning: "HAVE_ALLOC_PAIR" is not defined [-Wundef]
 #if !HAVE_ALLOC_PAIR
      ^~~~~~~~~~~~~~~
src/render.c:92:6: warning: "HAVE_EXTENDED_COLOR" is not defined [-Wundef]
 #if !HAVE_EXTENDED_COLOR
      ^~~~~~~~~~~~~~~~~~~
src/render.c: In function 'set_pair':
src/render.c:245:5: warning: "HAVE_ALLOC_PAIR" is not defined [-Wundef]
 #if HAVE_ALLOC_PAIR
     ^~~~~~~~~~~~~~~
src/render.c:248:7: warning: "HAVE_EXTENDED_COLOR" is not defined [-Wundef]
 #elif HAVE_EXTENDED_COLOR
       ^~~~~~~~~~~~~~~~~~~
src/render.c: In function 'set_color_pair_indirect':
src/render.c:275:5: warning: "HAVE_EXTENDED_COLOR" is not defined [-Wundef]
 #if HAVE_EXTENDED_COLOR
     ^~~~~~~~~~~~~~~~~~~
```


Perhaps using the following check instead for these `-Wundef` warnings?

```c
#if !defined(HAVE_ALLOC_PAIR)
```

Reply:
or just using `ifndef` and `ifdef` since we only check one by one.